src/vscode-dts/vscode.d.ts

```diff
@@ -19654,6 +19737,306 @@ declare module 'vscode' {
+		content: (LanguageModelTextPart | LanguageModelPromptTsxPart | unknown)[];
+
+		/**
+		 * @param callId The ID of the tool call.
+		 * @param content The content of the tool result.
+		 */
+		constructor(callId: string, content: (LanguageModelTextPart | LanguageModelPromptTsxPart | unknown)[]);
+	}
+
+	/**
+	 * A result returned from a tool invocation. If using `@vscode/prompt-tsx`, this result may be rendered using a `ToolResult`.
+	 */
+	export class LanguageModelToolResult {
+		/**
+		 * A list of tool result content parts. Includes `unknown` becauses this list may be extended with new content types in
+		 * the future.
+		 * @see {@link lm.invokeTool}.
+		 */
+		content: (LanguageModelTextPart | LanguageModelPromptTsxPart | unknown)[];
+
+		/**
+		 * Create a LanguageModelToolResult
+		 * @param content A list of tool result content parts
+		 */
+		constructor(content: (LanguageModelTextPart | LanguageModelPromptTsxPart | unknown)[]);
+	}
+
+	/**
+	 * A token that can be passed to {@link lm.invokeTool} when invoking a tool inside the context of handling a chat request.
+	 */
+	export type ChatParticipantToolToken = unknown;
+
+	/**
+	 * Options provided for tool invocation.
+	 */
+	export interface LanguageModelToolInvocationOptions<T> {
+		/**
+		 * When this tool is being invoked by a {@link ChatParticipant} within the context of a chat request, this token should be
+		 * passed from {@link ChatRequest.toolInvocationToken}. In that case, a progress bar will be automatically shown for the
+		 * tool invocation in the chat response view, and if the tool requires user confirmation, it will show up inline in the
+		 * chat view. If the tool is being invoked outside of a chat request, `undefined` should be passed instead.
+		 *
+		 * If a tool invokes another tool during its invocation, it can pass along the `toolInvocationToken` that it received.
+		 */
+		toolInvocationToken: ChatParticipantToolToken | undefined;
+
+		/**
+		 * The parameters with which to invoke the tool. The parameters must match the schema defined in
+		 * {@link LanguageModelToolInformation.parametersSchema}
+		 */
+		parameters: T;
+
+		/**
+		 * Options to hint at how many tokens the tool should return in its response, and enable the tool to count tokens
+		 * accurately.
+		 */
+		tokenizationOptions?: LanguageModelToolTokenizationOptions;
+	}
+
+	/**
+	 * Options related to tokenization for a tool invocation.
+	 */
+	export interface LanguageModelToolTokenizationOptions {
+		/**
+		 * If known, the maximum number of tokens the tool should emit in its result.
+		 */
+		tokenBudget: number;
+
+		/**
+		 * Count the number of tokens in a message using the model specific tokenizer-logic.
+		 * @param text A string.
+		 * @param token Optional cancellation token.  See {@link CancellationTokenSource} for how to create one.
+		 * @returns A thenable that resolves to the number of tokens.
+		 */
+		countTokens(text: string, token?: CancellationToken): Thenable<number>;
```

Comment:
https://github.com/microsoft/vscode/issues/231519: should this be `number | Thenable<number>`, or should the prompt-tsx version change to just `Promise<number>`? I'm thinking change the prompt-tsx version

Reply:
I prefer `Thenable<number>` because it is easy to consume